src/kudu/consensus/log_cache.cc

```diff
@@ -164,17 +197,182 @@ void LogCache::TruncateOpsAfterUnlocked(int64_t index) {
   next_sequential_op_index_ = index + 1;
 }
 
+Status LogCache::UncompressMsg(const ReplicateRefPtr& msg,
+                               faststring& buffer,
+                               std::unique_ptr<ReplicateMsg>* uncompressed_msg) {
+  const auto op_type = msg->get()->op_type();
+  const auto& payload = msg->get()->write_payload();
+  const auto compression_codec = payload.compression_codec();
+  const auto uncompressed_size = payload.uncompressed_size();
+  const auto compressed_size = payload.payload().size();
+
+  VLOG(2) << "Uncompressing messages before writing to log." <<
+             " opid: " << msg->get()->id().ShortDebugString() <<
+             " codec: " << compression_codec <<
+             " op_type: " << op_type <<
+             " compressed payload size: " << compressed_size <<
+             " uncompressed payload size: " << uncompressed_size;
+
+  // Resize buffer to hold uncompressed payload.
+  // TODO: needs perf testing and maybe implement streaming (un)compression
+  buffer.resize(uncompressed_size);
+  Slice compressed_slice(payload.payload().c_str(), compressed_size);
+
+  Status status;
+  const CompressionCodec* codec = codec_.load();
+  if (codec && codec->type() == compression_codec) {
+    status = codec->Uncompress(
+        compressed_slice, buffer.data(), uncompressed_size);
+  }
+  else {
+    // Either compression is not enabled on this instance OR this message uses a
+    // different compression codec. Get the right codec now
```

Comment:
the workflow for a message specific compression codec might be very difficult, because
we compress once and send it to all followers from LEADER. Is this so that we can change
the compression codec on the fly?

Reply:
Yes. If we change and/or disable compression, then the messages in flight could already be compressed and old messages in log-cache could already be compressed